File: q76-100/q79.py
# 2020.09.17
# Problem Statement:
# https://leetcode.com/problems/word-search/

class Solution:
    def existHelper(self, board, word, row, col, word_idx):      
        # current cell matches, check the next
        if word[word_idx] == board[row][col]:
            # set visit as true            
            self.board_visited[row][col] = True
            # found the whole word
            if word_idx == len(word)-1: return True            
        # current cell doesn't match, return to the previous
        else:
            word_idx = -1
            return False
        
        # search to the right
        # cannot go right any more
        if col>len(board[0])-2 or self.board_visited[row][col+1]:
            pass
        # can still go right, check the next cell
        else:
            right = self.existHelper(board, word, row, col+1, word_idx+1)
            if right:
                self.board_visited[row][col] = False
                return True
            
        # search to the left
        # cannot go left any more
        if col<1 or self.board_visited[row][col-1]:
            pass
        # can still go left, check the next cell
        else:
            left = self.existHelper(board, word, row, col-1, word_idx+1)
            if left:
                self.board_visited[row][col] = False
                return True

        # search to the top
        # cannot go top any more
        if row<1 or self.board_visited[row-1][col]:
            pass
        # can still go top, check the next cell
        else:
            top = self.existHelper(board, word, row-1, col, word_idx+1)
            if top:
                self.board_visited[row][col] = False
                return True

        # search to the bottom
        # cannot go bottom any more
        if row>len(board)-2 or self.board_visited[row+1][col]:
            pass
        # can still go bottom, check the next cell
        else:
            bottom = self.existHelper(board, word, row+1, col, word_idx+1)
            if bottom:
                self.board_visited[row][col] = False
                return True

        # this cell does not work, go back to the previous one
        self.board_visited[row][col] = False
        return False

        
    def exist(self, board: List[List[str]], word: str) -> bool:
        # initialize the board_visited which represented which cells are already visited
        self.board_visited = deepcopy(board)
        for i in range(0, len(board)):
            for j in range(0, len(board[0])):        
                self.board_visited[i][j] = False
        
        # check each cell as the start point
        for i in range(0, len(board)):
            for j in range(0, len(board[0])):
                if self.existHelper(board, word, i, j, 0):
                    return True
        
        return False

File: Amazon/q79.java
// 2022.05.20 midnight
// Problem Statement:
// https://leetcode.com/problems/word-search/

// idea: dfs
// can modify the board[x][y] to * to mark a visited node, instead of building a whole new structure
class Solution {
    boolean [][] visited;
    public boolean existHelper(char[][] board, String word, int x, int y) {
        if (word.length()==0) return true;
        if (word.charAt(0)!=board[x][y]) return false;
        if (word.length()==1 && word.charAt(0)==board[x][y]) return true;
        visited[x][y] = true;
        boolean temp = false;
        // top
        if (x>0 && !visited[x-1][y]) {
            temp = existHelper(board, word.substring(1), x-1, y);
            if (temp) return true;
        }
        // left
        if (y>0 && !visited[x][y-1]) {
            temp = existHelper(board, word.substring(1), x, y-1);
            if (temp) return true;
        }
        // bottom
        if (x<board.length-1 && !visited[x+1][y]) {
            temp = existHelper(board, word.substring(1), x+1, y);
            if (temp) return true;
        }
        // right
        if (y<board[0].length-1 && !visited[x][y+1]) {
            temp = existHelper(board, word.substring(1), x, y+1);
            if (temp) return true;
        }
        visited[x][y] = false;
        return false;
    }
    
    public boolean exist(char[][] board, String word) {
        // dfs
        boolean answer = false;
        visited = new boolean [board.length][board[0].length];
        for (int i=0; i<board.length; i++) {
            for (int j=0; j<board[0].length; j++) {
                answer = existHelper(board, word, i, j);
                if (answer) return true;
            }
        }
        return false;
    }
}